### DISCUSSION TITLE
Speed Controller track element

### DISCUSSION THREAD

**Discussion by ZehMatt** [OpenRCT2/OpenRCT2#19413] [Feature requests]
I'm a bit curious if it wouldn't make perhaps more sense to have a new track piece called "Speed Controller" which adds all those special things it can break or accelerate. I'm not sure if brakes should ever accelerate something, by definition brakes slow things down and that's about it, my question would be: any particular reason we can't just have a new element vs hacking new functionality into an existing element?

Originally posted by @ZehMatt in https://github.com/OpenRCT2/OpenRCT2/issues/18713#issuecomment-1431511663_

**Comment by ZehMatt** [comment]
To put a couple of properties of a speed controller to perspective:
```
- Minimum Speed: <number>          // Cars can't go below this speed.
- Maximum Speed: <number>          // Cars can't go above this speed.
- Target Speed:  <number>          // The actual desired speed.
- Acceleration:  <number>          // Amount of acceleration/deceleration towards target.
- Instantaneous: <bool>            // Set the target speed immediately.
```
Probably a couple of things I forgot.

This would also make some elements obsolete or rather can derive from the speed controller, block brakes would just set the max speed,
boosters would set acceleration and target speed, there are probably a couple more things which we could add.

**Comment by karst** [comment]
Also from the #18713 discussion. 

@X123M3-256 was working on reworked "Boosters" a little while back ([Discord message link](https://discord.com/channels/264137540670324737/691736850980667514/1027991431408791613)). 
This would include a reworked regular booster, a new vertical booster, and potentially a reworked sloped booster.

I am very doubtful it'd be necessary to have an extra element that's so similar to the booster, making that element redundant. Which is why I'd say if it's too much of an issue that it's called a booster, we might have to prompt on renaming that, instead of creating a fully new element out of it.

**Reply by ZehMatt** [reply]
It would not remove the current elements but those elements would rather derive from the speed controller logic, those boosters would just use functionality of the speed controller essentially, it all boils down to a couple of values that drives the behavior of all those elements.

**Comment by spacek531** [comment]
Currently blocking this is TrackElement structs are out of space. Last I checked, the available free space in the struct amounted to two unused Flags2 bits and two unused ColourScheme bits.

To get more space, Duncan has mentioned OpenRCT2 moving to an ECS model. This is something I am very interested in because I want more space for Track configurations. I have a [rudimentary implementation](https://github.com/spacek531/OpenRCT2/tree/extra-track-fields) but I did not make a PR for it because I was told the implementation was something the OpenRCT2 team wanted to handle internally.

**Comment by 733737** [comment]
Would this speed controller hypothetically accept negative values? IE currently you can have trains that traverse track in reverse using switchbacks with merged tracks that occupy the same tile. If we implemented a way to reverse trains using a special element, would the game glitch, fail to complete the ride or return to station, or would it cause the vehicles to explode?

**Reply by ZehMatt** [reply]
I see no reason why it shouldn't have negative acceleration or target speed. Why would vehicles explode going backwards?

**Reply by 733737** [reply]
I am simply wondering if we need to modify the ride station logic so that a train can arrive at station from the front. Vehicles can already depart station from the back and pass through the station...

**Reply by ocalhoun6** [reply]
Doesn't every shuttle loop type launched coaster already work this way, with trains arriving at the front of the station?

**Reply by 733737** [reply]
you misunderstand. I want the station to work in two directions. There were also mentions of the fact that IRL coasters can have track pieces that work similar to the reverser turntable on a log flume.

**Reply by ZehMatt** [reply]
Its a bit beyond the scope of this discussion about a single track element. This sounds more like an entire new operating mode.

**Reply by 733737** [reply]
sorry my bad, I just thought that this new element would introduce switching the train's direction, in which case, you would need to accomodate the game logic regarding multiple directions traversed on one ride, maybe it is a stretch to allow the station to be bidirectional, but there are IRL rides that have been built in OpenRCT2 that operate nearly simultaneously in two directions on the same track.
